### PR TITLE
Book and category test added

### PR DESCRIPTION
**Pull request by VlasenkoYevhenii**
image

### CODE REVIEW

File: src/test/java/com/example/bookstoreapplication/service/BookServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+                .setPrice(BigDecimal.valueOf(20))
+                .setDescription("A romantic novel")
+                .setCoverImage("https://example.com/book2-cover-image.jpg");
+        Book book1984 = new Book()
+                .setTitle("1984")
+                .setAuthor("George Orwell")
+                .setIsbn("9780451524935")
+                .setPrice(BigDecimal.valueOf(15))
+                .setDescription("A dystopian social science fiction novel")
+                .setCoverImage("https://example.com/book3-cover-image.jpg");
```
Comment (DmytroHadiuchko) [inline src/test/java/com/example/bookstoreapplication/service/BookServiceImplTest.java:84]: it seems to me that there is a lot of hardcoding here, maybe it's would be better to move it as constant

File: src/test/java/com/example/bookstoreapplication/service/CategoryServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        category.setDescription("Fictional books");
+
+        categoryDto = new CategoryDto();
+        categoryDto.setId(1L);
+        categoryDto.setName("Fiction");
+        categoryDto.setDescription("Fictional books");
+
+        categoryRequestDto = new CategoryRequestDto();
+        categoryRequestDto.setName("Fiction");
+        categoryRequestDto.setDescription("Fictional books");
```
Comment (DmytroHadiuchko) [inline src/test/java/com/example/bookstoreapplication/service/CategoryServiceImplTest.java:62]: you reuse name and description, maybe would be better move it to a constant

**Review by DmytroHadiuchko — APPROVED** [review]
It looks good, but it seems to me that it too much hardcoding
you have the same situation with IDs, check it

File: .env [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+JWT_SECRET=vhc498ht4mwt94t94t9gt3h9g3gh3795xgh93h5g938gh3ughreg
+JWT_EXPIRATION=300000000
+ARRAY_OF_ENDPOINTS=/api/auth/**,/swagger-ui/**,/v3/api-docs/**
+
+MYSQL_LOCAL_PORT=3306
+MYSQL_DOCKER_PORT=3307
+SPRING_LOCAL_PORT=8080
+SPRING_DOCKER_PORT=8080
+DEBUG_PORT=5005
```
Comment (karma-o) [inline .env:18]: better to not push that file

File: src/test/java/com/example/bookstoreapplication/service/BookServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+                .setDescription(book1984.getDescription())
+                .setCoverImage(book1984.getCoverImage());
+
+        books = List.of(greatGatsby, prideAndPrejudice, book1984);
+        bookDtos = List.of(greatGatsbyDto, prideAndPrejudiceDto, book1984Dto);
+    }
+
+    @Test
+    @DisplayName("Verify save() method with correct requestDto")
+    void saveBook_ValidBookRequest_ReturnBookDto() {
```
Comment (karma-o) [inline src/test/java/com/example/bookstoreapplication/service/BookServiceImplTest.java:114]: can we add some negative test cases for save method?

File: src/test/java/com/example/bookstoreapplication/service/BookServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+                .setPrice(theHobbit.getPrice());
+
+        when(bookMapper.toModel(theHobbitRequestDto)).thenReturn(theHobbit);
+        when(bookRepository.save(theHobbit)).thenReturn(theHobbit);
+        when(bookMapper.toDto(theHobbit)).thenReturn(theHobbitDto);
+
+        BookDto actual = bookServiceImpl.save(theHobbitRequestDto);
+        Assertions.assertNotNull(actual);
+        Assertions.assertEquals(theHobbitDto, actual);
+        verify(bookRepository, times(ONE_TIME)).save(theHobbit);
```
Comment (karma-o) [inline src/test/java/com/example/bookstoreapplication/service/BookServiceImplTest.java:138]:
```suggestion
        verify(bookRepository, once()).save(theHobbit);
```

File: src/test/java/com/example/bookstoreapplication/service/CategoryServiceImplTest.java (hunk 2 above)
Comment (karma-o) [reply src/test/java/com/example/bookstoreapplication/service/CategoryServiceImplTest.java:62]: agree

File: src/test/java/com/example/bookstoreapplication/service/CategoryServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        when(categoryMapper.toDto(category)).thenReturn(categoryDto);
+
+        CategoryDto actual = categoryService.update(1L, categoryRequestDto);
+        assertNotNull(actual);
+        assertThat(actual).isEqualTo(categoryDto);
+
+        verify(categoryMapper, times(1)).toModel(categoryRequestDto);
+        verify(categoryRepository, times(1)).save(category);
+        verify(categoryMapper, times(1)).toDto(category);
+    }
```
Comment (karma-o) [inline src/test/java/com/example/bookstoreapplication/service/CategoryServiceImplTest.java:138]: again, can we add some negative cases, where service methods can throw us an exception as a response to some invalid input

File: src/test/java/com/example/bookstoreapplication/service/BookServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+                .findAll(PageRequest.of(ZERO_PAGE_NUMBER, DEFAULT_PAGE_SIZE));
+        Assertions.assertEquals(3, actual.size());
+        Assertions.assertEquals(expected, actual);
+        verify(bookRepository, times(ONE_TIME))
+                .findAll(PageRequest.of(ZERO_PAGE_NUMBER, DEFAULT_PAGE_SIZE));
+    }
+
+    @Test
+    @DisplayName("Verify getBookDtoById() method with correct bookId")
+    void findBookById_ValidBookId_ReturnBookDto() {
```
Comment (karma-o) [inline src/test/java/com/example/bookstoreapplication/service/BookServiceImplTest.java:167]: where we have a `... by id` we can mock repo method that returns empty Optional and verify the behavior of our method in such cases, same for category tests

**Review by karma-o — CHANGES_REQUESTED** [review]
Missing integrational API controller tests with the use of test containers, let's add them for book and category controllers, and check other comments as well

**Comment by VlasenkoYevhenii** [comment]
Screenshot_3

**Review by karma-o — APPROVED** [review]
Looks very good!

**Pull request merged**